Major API changes and updated README

onraj55555 opened the pull request:
API changes:
- comments in right format
- temporary removed dependency (and so functions using it) wasi:io
- added extra functionality to output pins
- added back constructors

README changes:
- Added phase of the proposal

onraj55555 committed:
rewrite: api, not finished

onraj55555 committed:
update: major api changes

onraj55555 committed:
update: readme

sunfishcode commented on wit/gpio.wit:
```diff
@@ -32,259 +52,385 @@ interface general {
         set-active: func() -> result<_, gpio-error>;
-        // Sets the pin to the inactive state
+        /// Sets the pin to the inactive state
         set-inactive: func() -> result<_, gpio-error>;
```
In terms of an embedded-hal implementation, these new functions appear to correspond to [`StatefulOutputPin`](https://docs.rs/embedded-hal/latest/embedded_hal/digital/trait.StatefulOutputPin.html) rather than just [`OutputPin`](https://docs.rs/embedded-hal/latest/embedded_hal/digital/trait.OutputPin.html). Do you know if this significantly impacts what kinds of devices this API could be implemented on?

onraj55555 replied:
Comments on the StatefulOutPin comments:
You're right about that. When making this API I took inspiration from [rppal](https://crates.io/crates/rppal) which also supports this functionality. I took some distance from embedded-hal because the naming convention uses high and low for pin states while this API used active and inactive which allows for a more general usage (configuring it to be active low for example). I have no idea yet what the impact it could have but by looking at this post ([here](https://forum.arduino.cc/t/how-to-read-the-state-of-an-output-pin/100490)), Arduino microcontroller output pins are stateful. The RP2040 also supports this ([here](https://forums.raspberrypi.com/viewtopic.php?t=353513)). ESP32 seems to have a bit more of a problem with this ([here](https://esp32.com/viewtopic.php?t=5793)) but this is quite an old post. I will add a new resource for StatefulOutputPins so this API will be compatible with all devices.

Comments on removing wasi:io:
After experimenting with implementing the API I thought that the functions which returned a pollable were on the complex side because not all microcontrollers support concurrent executions which was required for these functions. Looking at Arduino again, only certain pins can be used for interrupts ([here](https://reference.arduino.cc/reference/en/language/functions/external-interrupts/attachinterrupt/)). The biggest concern I had was that these functions would not be usable in most implementations but I could be wrong in that. It could give performance improvements by reducing the overhead when reading the state from an input pin.

Without the pollable functions this is my vision on calling read: guest calls read -> host implementations for read gets called -> board specific function gets called.
This could cause some latency when this is called in a fast loop.

With pollables this could be possible:
- guest calls read pollable -> host sets up a task for that specific state change and returns
- guest calls the pollable -> the host just needs to check if the pollable is triggered and that happens in the background

To see if this has a major performance impact, I will need to perform some tests.

Thanks for the comments!

sunfishcode reviewed (commented):
Could you comment on why wasi:io is removed?

onraj55555 committed:
Updated GPIO API to be cleaner by only using resources and methods for
the main API. Delay is transformed to be resource based but will be
replaced by wasi:clocks in the future.

onraj55555 committed:
forgot some things

onraj55555 committed:
Update gpio.wit

result type for get-config function

onraj55555 committed:
extra error types

onraj55555 committed:
- removed watch functions from digital and analog IO pins because
  implementing them would cause a lot of issues with the automatic mode
change
- added set-pin-mode function to digital and analog IO pins

onraj55555 committed:
Merge branch 'main' of github.com:onraj55555/wasi-gpio
